bentoml/metrics.py

```diff
@@ -0,0 +1,94 @@
+from __future__ import annotations
+
+import typing as t
+import logging
+from abc import ABC
+from abc import abstractmethod
+from typing import TYPE_CHECKING
+
+from simple_di import inject
+from simple_di import Provide
+
+logger = logging.getLogger(__name__)
+
+from ._internal.configuration.containers import BentoMLContainer
+
+if TYPE_CHECKING:
+    from bentoml._internal.server.metrics.prometheus import PrometheusClient
+
+
+class BaseMetric(ABC):
```

Comment:
I don't think `BaseMetric` makes sense here, since it is not really a Base class, rather than a meta class.

Reply:
Do you have any thoughts about what I can do to finalize the design?

Reply:
up to you

Reply:
I am not sure that the benefits of introducing meta-classes here outweighs the drawbacks, especially when considering code readability.

I think as it is right now, there is a good balance between functionality and simplicity. However, with a little bit of guidance I don't mind converting this into a Meta Class. 

lmk.

Reply:
it is not essentially just a metaclass, because metaclass is quite different in for this case specifically. BaseMetric doesn't make sense to me since it collides and makes people confuse with `prometheus_client.Metrics`

Reply:
Do you think that making `BaseMetric` a "protected" class and providing a docstring string would be enough to avoid the clash?

Reply:
Just call it `_MetricsMeta`

Reply:
would you mind if I prefix it with 'I' as in `_IMetricsMeta`? 😜